File: tests/hardware/ucitime.py
#!/usr/bin/env python3
"""
How long is one UCI round trip, in frames?

Runs tests/hardware/ucitime.prg on a real Ultimate and reads its results back by
DMA. The C64 measures; this only divides and prints.

    python3 ucitime.py --host 192.168.1.62

Every number the program publishes is in CIA cycles off one chained 32-bit
counter, and one of those numbers is the length of a raster frame measured on
that same counter. So "frames per call" is a ratio of two figures from one
clock: it does not care whether the machine is PAL or NTSC, what the CIA is
really clocked at on an Ultimate 64, or whether the CPU is in turbo. The turbo
register is reported so the answer can be read in context.

This touches no settings. Wrap it in tools/u64_settings.py if the command
interface might be disabled, which on the bench machine it usually is - the
Makefile's time-run target does exactly that.

Part of the Ultimate SDK. SPDX-License-Identifier: MIT
"""

import argparse
import math
import os
import sys
import time
import urllib.error

REPO = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(REPO, "tools"))

from u64_settings import Ultimate  # noqa: E402

from screen import decode_screen, SCREEN_ADDR, SCREEN_LEN  # noqa: E402

# The result block ucitime.prg publishes into the cassette buffer. Layout is
# documented beside publish() in ucitime.c; keep the two in step.
RESULT_ADDR = 0x033C
RESULT_LEN = 59
RESULT_MAGIC = b"UCIM"
RESULT_DONE = 0xA5
MAX_SLOTS = 8

# Slot names, in the order ucitime.c fills them. A slot the program measured but
# this list does not name still reports, under its number - so adding a
# measurement to the C side and forgetting this list degrades rather than lies.
NAMES = [
    "identify",
    "echo",
    "echo (no reply)",
    "get-palette",
    "set-palette-color",
    "set-palette-color (no reply)",
    "set-palette (all 16)",
    "set-palette (no reply)",
]

def error_names():
    """ULTIMATE_* codes, read out of the generated header rather than copied.

    A transcribed copy of this table was wrong within five minutes of being
    written, which is the same lesson the argument shapes taught in
    docs/handover.md section 2: the generated file is the source of truth, so
    read it.
    """
    out = {}
    path = os.path.join(REPO, "include", "uci_protocol.h")
    with open(path) as fh:
        for line in fh:
            parts = line.split()
            if len(parts) >= 3 and parts[0] == "#define" and \
                    (parts[1] == "ULTIMATE_OK" or parts[1].startswith("ULTIMATE_ERR_")):
                name = parts[1].replace("ULTIMATE_ERR_", "").replace("ULTIMATE_", "")
                out[int(parts[2])] = name.lower().replace("_", " ")
    return out


ERRORS = error_names()


def u32(block, at):
    return (block[at] | (block[at + 1] << 8) |
            (block[at + 2] << 16) | (block[at + 3] << 24))


class Result:
    def __init__(self, block):
        self.valid = block[0:4] == RESULT_MAGIC
        if not self.valid:
            return
        self.format = block[4]
        self.slots = block[5]
        self.pal = block[6]
        self.iters = block[7]
        self.turbo = block[8]
        self.frame = u32(block, 10)
        self.overhead = u32(block, 14)
        self.cycles = [u32(block, 18 + 4 * i) for i in range(MAX_SLOTS)]
        self.errors = list(block[50:50 + MAX_SLOTS])
        self.done = block[58] == RESULT_DONE


def run_once(u, prg, settle=6.0, poll_timeout=60.0):
    """Boot the program and wait for it to publish. Same shape as hwtest.py."""
    u.writemem(RESULT_ADDR, bytes(RESULT_LEN))
    u.run_prg(prg)

    deadline = time.time() + poll_timeout
    time.sleep(settle)
    result = None
    while time.time() < deadline:
        result = Result(u.readmem(RESULT_ADDR, RESULT_LEN))
        if result.valid and result.done:
            break
        time.sleep(1.0)
    screen = decode_screen(u.readmem(SCREEN_ADDR, SCREEN_LEN))
    return result, screen


# What a frame is on a stock machine, printed beside the measured figure. The
# measurement is the one used for every ratio below; this is only here so that a
# broken measurement is obvious instead of merely plausible - which is exactly
# how the first version's missing raster sync was caught.
TEXTBOOK_FRAME = {1: ("PAL", 63 * 312), 0: ("NTSC", 65 * 263)}


def report(result):
    kind, expected = TEXTBOOK_FRAME.get(result.pal, ("unknown", 0))
    print("# frame        = %d cycles, measured (%s; a stock one is %d)"
          % (result.frame, kind, expected))
    print("# turbo reg    = $%02X (%s)"
          % (result.turbo,
             "turbo unavailable" if result.turbo == 0xFF
             else "turbo on, speed index %d" % (result.turbo & 0x0F)))
    print("# timer cost   = %d cycles, subtracted from every measurement"
          % result.overhead)
    print("# %d calls averaged per measurement" % result.iters)
    print()
    print("%-30s %9s %8s %10s" % ("command", "cycles", "frames", "per frame"))

    fastest = None
    for i in range(result.slots):
        name = NAMES[i] if i < len(NAMES) else "measurement %d" % i
        err = result.errors[i]
        if err != 0 or result.cycles[i] == 0:
            print("%-30s %9s %8s %10s   %s"
                  % (name, "-", "-", "-", ERRORS.get(err, "error %d" % err)))
            continue
        per_call = result.cycles[i] / result.iters
        frames = per_call / result.frame
        print("%-30s %9d %8.2f %10.1f"
              % (name, round(per_call), frames, 1.0 / frames))
        if fastest is None or frames < fastest[1]:
            fastest = (name, frames)

    if fastest is None:
        print("\nnothing measured.")
        return

    # The question this was written to answer: a colour cycle is every colour
    # moving at once, so the whole-palette write is the number that decides
    # whether the boing ball can cycle per frame. See docs/handover-next.md.
    rotation = NAMES.index("set-palette (all 16)")
    if rotation < result.slots and result.errors[rotation] == 0 \
            and result.cycles[rotation]:
        per_call = result.cycles[rotation] / result.iters
        frames = per_call / result.frame
        print()
        print("a whole-palette rotation - all 16 colours, one command - costs")
        print("%.2f frames (%.1f per frame)." % (frames, 1.0 / frames))

    name, frames = fastest
    print()
    print("cheapest round trip: %s, %.2f frames (%.1f per frame)"
          % (name, frames, 1.0 / frames))
    if frames <= 0.5:
        print("verdict: several fit in a frame. cycle the palette directly.")
    elif frames <= 1.0:
        print("verdict: one fits in a frame, with nothing to spare. cycle every")
        print("         frame only if the demo has cycles left over.")
    else:
        print("verdict: a round trip costs more than a frame. cycle every %d"
              % math.ceil(frames))
        print("         frames, or drive the C64's own colour registers and use")
        print("         the Ultimate's palette to change what they mean.")


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--host", required=True, help="hostname or IP of the Ultimate")
    ap.add_argument("--port", type=int, default=80)
    ap.add_argument("--prg", default="ucitime.prg")
    ap.add_argument("--verbose", action="store_true", help="print the C64 screen")
    args = ap.parse_args()

    u = Ultimate(args.host, args.port)
    try:
        info = u.info()
    except (urllib.error.URLError, OSError) as exc:
        print("cannot reach the Ultimate at %s: %s" % (args.host, exc))
        return 2

    print("# %s, firmware %s (fpga %s, core %s)"
          % (info.get("product"), info.get("firmware_version"),
             info.get("fpga_version"), info.get("core_version")))

    result, screen = run_once(u, args.prg)

    if result is None or not result.valid or not result.done:
        print("no result block: the program did not reach the end.")
        for line in screen:
            print("| %s" % line)
        return 1

    report(result)

    if args.verbose:
        print()
        for line in screen:
            print("| %s" % line)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/hardware/ucitime.c
/*
 * ucitime.c - how long is one UCI round trip, in frames?
 *
 * docs/handover-next.md asks for one number, because it decides the shape of
 * anything that wants to talk to the Ultimate while a demo is running. Colour
 * cycling wants a palette write every frame or two; nobody had measured whether
 * a round trip fits in a frame at all.
 *
 * Every UCI command is a handshake: write the bytes, push, poll until BUSY
 * clears, read the status. That poll is the cost, and it is the Ultimate's
 * turnaround, not the C64's - so it can only be measured on real hardware.
 *
 * **The frame length is measured, not assumed.** CIA #2's two timers are
 * chained into a 32-bit cycle counter, and the same counter times ten raster
 * frames. Every result is then a ratio of two numbers off one clock, which is
 * what makes it immune to PAL versus NTSC, to whatever the CIA is really
 * clocked at on an Ultimate 64, and to the machine sitting in turbo. The turbo
 * register is published alongside so the report can say which it was.
 *
 * Build:  make ucitime.prg
 * Run:    make time-run U64_HOST=<ip>, which reads the results back by DMA.
 *
 * SPDX-License-Identifier: MIT
 */

#include <stdio.h>
#include <string.h>
#include <peekpoke.h>
#include <ultimate.h>

/* ---- CIA #2, chained into a 32-bit cycle counter --------------------------
 *
 * CIA #2 is the free one: the KERNAL drives its timers only for RS232, and
 * nothing here touches $DD00, so the VIC bank and the serial bus are unharmed.
 * Timer A counts phi2, timer B counts timer A's underflows.
 */
#define CIA2_TA_LO 0xDD04
#define CIA2_TA_HI 0xDD05
#define CIA2_TB_LO 0xDD06
#define CIA2_TB_HI 0xDD07
#define CIA2_ICR   0xDD0D
#define CIA2_CRA   0xDD0E
#define CIA2_CRB   0xDD0F

#define CRA_RUN    0x11            /* force load, count phi2, start */
#define CRB_RUN    0x51            /* force load, count timer A underflows, start */

#define VIC_RASTER 0xD012
#define VIC_TURBO  0xD031          /* $FF when the Ultimate's turbo is not available */
#define PAL_FLAG   0x02A6          /* KERNAL: 1 = PAL, 0 = NTSC */

static void timer_start(void)
{
    POKE(CIA2_CRA, 0x00);
    POKE(CIA2_CRB, 0x00);
    POKE(CIA2_ICR, 0x7F);          /* no NMI from either timer, whatever ran before */
    (void)PEEK(CIA2_ICR);          /* reading clears the latched flags */
    POKE(CIA2_TA_LO, 0xFF);
    POKE(CIA2_TA_HI, 0xFF);
    POKE(CIA2_TB_LO, 0xFF);
    POKE(CIA2_TB_HI, 0xFF);
    POKE(CIA2_CRB, CRB_RUN);       /* B first: it must already be counting when A wraps */
    POKE(CIA2_CRA, CRA_RUN);
}

/*
 * Both timers count down from $FFFF and reload on underflow, so a timer A
 * period is $FFFF + 1 cycles. Stop before reading: a rollover between the two
 * reads would otherwise be worth 65536 cycles of nonsense.
 */
static uint32_t timer_stop(void)
{
    uint16_t a, b;

    POKE(CIA2_CRA, 0x00);
    POKE(CIA2_CRB, 0x00);
    a = PEEK(CIA2_TA_LO) | ((uint16_t)PEEK(CIA2_TA_HI) << 8);
    b = PEEK(CIA2_TB_LO) | ((uint16_t)PEEK(CIA2_TB_HI) << 8);
    return ((uint32_t)(0xFFFFu - b) << 16) | (uint32_t)(uint16_t)(0xFFFFu - a);
}

/* ------------------------------------------------------------------------ */

#define MAX_SLOTS 8
#define ITERS     8                /* per measurement, to average out badline jitter */

static uint32_t cycles[MAX_SLOTS];
static uint8_t  errors[MAX_SLOTS];
static uint8_t  slots;

static uint32_t overhead;          /* timer start + stop, charged once per measurement */
static uint32_t frame_cycles;      /* one raster frame, on the same clock */

/*
 * Interrupts off for the whole burst. The KERNAL's own IRQ is not part of a
 * round trip, and leaving it in would measure the jiffy clock as well. The
 * screen stays on: a demo's screen is on, so the VIC's stolen cycles belong in
 * the number.
 */
static uint32_t time_exec(uci_request *req, uint8_t slot)
{
    uint8_t i;
    uint32_t total;

    __asm__("sei");
    timer_start();
    for (i = 0; i < ITERS; ++i)
        errors[slot] = uci_exec(req);
    total = timer_stop();
    __asm__("cli");

    return total > overhead ? total - overhead : 0;
}

/*
 * Ten frames on the same counter. This is the calibration: everything else is
 * reported as a ratio against it, so nothing here has to know how long a frame
 * is supposed to be.
 *
 * Raster line 250 exists once per frame on PAL and on NTSC alike - $D012 is the
 * low eight bits of a line number that never reaches 506 - so one wait-for-250
 * plus one wait-to-leave-250 is exactly one frame.
 *
 * Sync before starting the timer, not after. Starting it first puts a partial
 * frame in front of the ten: the first measurement read 16245 cycles where an
 * NTSC frame is 17095, which is 9.5 frames divided by 10 and looked like a
 * plausible number rather than an obviously broken one.
 */
static uint32_t time_ten_frames(void)
{
    uint8_t i;
    uint32_t total;

    __asm__("sei");
    while (PEEK(VIC_RASTER) != 250)
        ;
    while (PEEK(VIC_RASTER) == 250)
        ;
    timer_start();
    for (i = 0; i < 10; ++i) {
        while (PEEK(VIC_RASTER) != 250)
            ;
        while (PEEK(VIC_RASTER) == 250)
            ;
    }
    total = timer_stop();
    __asm__("cli");

    return (total > overhead ? total - overhead : 0) / 10;
}

/* ------------------------------------------------------------------------ */

/*
 * Machine-readable results, in the cassette buffer, exactly as ucitest.c does
 * it: 192 bytes neither BASIC nor cc65 touches, cleared by the KERNAL's reset
 * so a stale block cannot be read as a fresh one.
 *
 *   +0  magic "UCIM"      +8  $D031             +18 8 x 4 cycle totals
 *   +4  format version    +9  reserved          +50 8 x 1 error code
 *   +5  measurements      +10 frame cycles      +58 $A5 once main() is done
 *   +6  $02A6, 1 = PAL    +14 timer overhead
 *   +7  iterations
 */
#define RESULT_BLOCK  ((uint8_t *)0x033C)
#define RESULT_FORMAT 1
#define RESULT_DONE   0xA5

static void put32(uint8_t *at, uint32_t v)
{
    at[0] = (uint8_t)v;
    at[1] = (uint8_t)(v >> 8);
    at[2] = (uint8_t)(v >> 16);
    at[3] = (uint8_t)(v >> 24);
}

static void publish(void)
{
    uint8_t i;

    RESULT_BLOCK[0] = 0x55;             /* 'U' - written as bytes, not a  */
    RESULT_BLOCK[1] = 0x43;             /* 'C'   literal, so no charmap   */
    RESULT_BLOCK[2] = 0x49;             /* 'I'   can rewrite it           */
    RESULT_BLOCK[3] = 0x4D;             /* 'M' */
    RESULT_BLOCK[4] = RESULT_FORMAT;
    RESULT_BLOCK[5] = slots;
    RESULT_BLOCK[6] = PEEK(PAL_FLAG);
    RESULT_BLOCK[7] = ITERS;
    RESULT_BLOCK[8] = PEEK(VIC_TURBO);
    RESULT_BLOCK[9] = 0;
    put32(RESULT_BLOCK + 10, frame_cycles);
    put32(RESULT_BLOCK + 14, overhead);
    for (i = 0; i < MAX_SLOTS; ++i) {
        put32(RESULT_BLOCK + 18 + 4 * i, cycles[i]);
        RESULT_BLOCK[50 + i] = errors[i];
    }
    RESULT_BLOCK[58] = RESULT_DONE;
}

/* ------------------------------------------------------------------------ */

static uci_request req;
static uint8_t     scratch[64];
static uint8_t     palette[UCI_PALETTE_BYTES];
static uint8_t     echo_args[4] = { 0xDE, 0xAD, 0xBE, 0xEF };
static uint8_t     color_args[4];

/* Every measurement is the same three steps, so they are one function. */
static void record(const char *name, uint8_t slot)
{
    cycles[slot] = time_exec(&req, slot);
    if (slot >= slots)
        slots = slot + 1;
    printf("%-14s %6lu %s\n", name, cycles[slot] / ITERS,
           errors[slot] == ULTIMATE_OK ? "" : "err");
}

static void skip(const char *name, uint8_t slot, uint8_t err)
{
    errors[slot] = err;
    cycles[slot] = 0;
    if (slot >= slots)
        slots = slot + 1;
    printf("%-14s      - skipped\n", name);
}

int main(void)
{
    uint8_t err;
    uint8_t have_palette;

    printf("uci round trip, in cycles\n\n");

    if (ultimate_init() != ULTIMATE_OK) {
        printf("no ultimate. enable the command\n");
        printf("interface in its settings menu.\n");
        publish();
        return 1;
    }

    /* The zero baseline: what the timer costs to start and stop. */
    __asm__("sei");
    timer_start();
    overhead = timer_stop();
    __asm__("cli");

    frame_cycles = time_ten_frames();
    printf("frame          %6lu cycles\n", frame_cycles);
    printf("timer overhead %6lu\n\n", overhead);

    /* 0: identify. A short reply, and the command every program starts with. */
    memset(&req, 0, sizeof(req));
    req.target  = UCI_TARGET_DOS1;
    req.command = UCI_CMD_IDENTIFY;
    req.data    = scratch;
    req.datamax = sizeof(scratch);
    record("identify", 0);

    /* 1: echo, four bytes out and six back - about as small as a command gets. */
    memset(&req, 0, sizeof(req));
    req.target  = UCI_TARGET_DOS1;
    req.command = DOS_CMD_ECHO;
    req.args    = echo_args;
    req.arglen  = sizeof(echo_args);
    req.data    = scratch;
    req.datamax = sizeof(scratch);
    record("echo", 1);

    /* 2: the same command with the reply thrown away, which is the fire-and-
     * forget path a demo would actually use. */
    memset(&req, 0, sizeof(req));
    req.target  = UCI_TARGET_DOS1 | UCI_TARGET_NO_REPLY;
    req.command = DOS_CMD_ECHO;
    req.args    = echo_args;
    req.arglen  = sizeof(echo_args);
    record("echo-noreply", 2);

    /*
     * The palette commands are firmware 3.15-and-newer. Probing with the read
     * first does double duty: it says whether to measure them at all, and it
     * hands back the colour to write, so the write below sets colour 0 to the
     * value it already has. Nothing on the screen changes, which is what keeps
     * this out of the destructive bucket.
     */
    memset(&req, 0, sizeof(req));
    req.target  = UCI_TARGET_CONTROL;
    req.command = CTRL_CMD_GET_PALETTE;
    req.data    = palette;
    req.datamax = sizeof(palette);
    err = uci_exec(&req);
    have_palette = (err == ULTIMATE_OK && req.datalen == UCI_PALETTE_BYTES);

    if (have_palette) {
        /* 3: read the whole palette - 48 bytes back. */
        record("get-palette", 3);

        color_args[0] = 0;                  /* colour 0 */
        color_args[1] = palette[0];
        color_args[2] = palette[1];
        color_args[3] = palette[2];

        /* 4: one colour, six bytes on the wire. The cheapest command there is. */
        memset(&req, 0, sizeof(req));
        req.target  = UCI_TARGET_CONTROL;
        req.command = CTRL_CMD_SET_PALETTE_COLOR;
        req.args    = color_args;
        req.arglen  = sizeof(color_args);
        record("set-color", 4);

        /* 5: and the same, fire and forget. */
        memset(&req, 0, sizeof(req));
        req.target  = UCI_TARGET_CONTROL | UCI_TARGET_NO_REPLY;
        req.command = CTRL_CMD_SET_PALETTE_COLOR;
        req.args    = color_args;
        req.arglen  = sizeof(color_args);
        record("set-color-nr", 5);

        /*
         * 6 and 7: all sixteen colours in one command. This is the one a colour
         * cycling demo actually wants - a rotation is every colour moving at
         * once - and sixteen separate SET_PALETTE_COLORs is the alternative it
         * is being compared against. Writing back exactly what was read leaves
         * the screen alone.
         */
        memset(&req, 0, sizeof(req));
        req.target  = UCI_TARGET_CONTROL;
        req.command = CTRL_CMD_SET_PALETTE;
        req.args    = palette;
        req.arglen  = sizeof(palette);
        record("set-palette", 6);

        memset(&req, 0, sizeof(req));
        req.target  = UCI_TARGET_CONTROL | UCI_TARGET_NO_REPLY;
        req.command = CTRL_CMD_SET_PALETTE;
        req.args    = palette;
        req.arglen  = sizeof(palette);
        record("set-palette-nr", 7);
    } else {
        skip("get-palette", 3, err);
        skip("set-color", 4, err);
        skip("set-color-nr", 5, err);
        skip("set-palette", 6, err);
        skip("set-palette-nr", 7, err);
    }

    printf("\nper call, %u calls each\n", ITERS);
    publish();
    return 0;
}
